spec/std/string_spec.cr

```diff
@@ -1851,16 +1851,34 @@ describe "String" do
     "ぜんぶ".chars.should eq(['ぜ', 'ん', 'ぶ'])
   end
 
-  it "allows creating a string with zeros" do
-    p = Pointer(UInt8).malloc(3)
-    p[0] = 'a'.ord.to_u8
-    p[1] = '\0'.ord.to_u8
-    p[2] = 'b'.ord.to_u8
-    s = String.new(p, 3)
-    s[0].should eq('a')
-    s[1].should eq('\0')
-    s[2].should eq('b')
-    s.bytesize.should eq(3)
+  describe "creating from a pointer" do
+    it "allows creating a string with zeros" do
+      p = Pointer(UInt8).malloc(3)
+      p[0] = 'a'.ord.to_u8
+      p[1] = '\0'.ord.to_u8
+      p[2] = 'b'.ord.to_u8
+      s = String.new(p, 3)
+      s[0].should eq('a')
+      s[1].should eq('\0')
+      s[2].should eq('b')
+      s.bytesize.should eq(3)
+    end
+
+    it "raises an exception when creating a string with a null pointer and no size" do
+      expect_raises ArgumentError do
+        String.new(Pointer(UInt8).null)
+      end
+    end
+
+    it "raises when creating from a null pointer with a nonzero size" do
+      expect_raises ArgumentError do
+        String.new(Pointer(UInt8).null, 3)
+      end
+    end
+
+    it "returns an empty string when creating from a null pointer with size 0" do
+      String.new(Pointer(UInt8).null, 0).should eq ""
```

Comment:
I'm fine with this behavior, but it might warrant some discussion.

Reply:
@jhass Why? This PR doesn't change the way it works right now.

Reply:
Previously null pointers were just not considered by the implementation. That it doesn't change  behavior for this case is more of an implementation detail than design. This considers how to deal with null pointers from a design perspective, so the consideration to change this is within scope in my opinion.

Reply:
I'm surprised there were no specs about pointers with 0 length because it seemed to be an intentional choice in the code.

Reply:
I think I would prefer `String.new(Pointer(UInt8).null, 0)` to raise. As raised by @jhass null pointers here were probably not considered before. It think it will also be more consistent with the behavior of `String.new(Pointer(UInt8).null)`.

The optimization that lead to the `bytesize == 0` check is about reusing the empty string literal embeded in the binary.

@jgaskins I know it's been a while, let me know if you prefer to hand of these changes.

Reply:
@bcardiff I'm happy to turn it over to someone that understands the implications better than I do. I mainly wanted to see if there was a better way of handling `char *s = 0` from a C function since someone I was chatting with about Crystal ran into it.

Reply:
@jhass @asterite Do you agree then to make `String.new(Pointer(UInt8).null, 0)` raise? If so, I can finish this up.

Thanks @jgaskins . The intention was to fail in a more controlled way in case of `char *s = 0`, right? Is not that that person was expecting `s` to be an empty string, right?

Reply:
Whichever makes the most sense. I think either case is better than a segfault. I originally had the same perspective @asterite did in [this comment](https://github.com/crystal-lang/crystal/pull/9653#issuecomment-679171410), but your response to it makes a lot of sense. And since any time we're dealing with pointers, we're generally also dealing with C code, I feel like making the distinction between an empty string and a null pointer seems like a reasonable requirement.

Reply:
I feel okay either way. We can see if somebody runs into trouble with the changed behavior. It seems reasonable to me that something would return a null pointer and a zero length in an error or "not found" case, but have no example at hand to back it up.